AirCasting/CoreData/PersistenceController+Database.swift

```diff
@@ -0,0 +1,141 @@
+                }
+            }
+            do {
+                try context.save()
+                completion?(nil)
+            } catch {
+                completion?(error)
+            }
+        }
+    }
+}
+
+extension PersistenceController: SessionRemovable {
+    func removeSessions(where constraint: Database.Constraint, completion: ((Error?) -> Void)?) {
+        let context = self.newBackgroundContext()
+        context.perform {
+            let request: NSFetchRequest<SessionEntity> = SessionEntity.fetchRequest()
+            if case .predicate(let predicate) = constraint {
+                request.predicate = predicate
+            }
+            do {
+                let fetchedEntities = try context.fetch(request)
+                fetchedEntities.forEach(context.delete(_:))
+                try context.save()
+                completion?(nil)
+            } catch {
+                completion?(error)
+            }
+        }
+    }
+}
+
+extension Database.Session {
+    init(coreDataEntity: SessionEntity) {
+        self.init(uuid: coreDataEntity.uuid,
+                  type: coreDataEntity.type,
+                  name: coreDataEntity.name,
+                  deviceType: coreDataEntity.deviceType,
+                  location: coreDataEntity.location,
+                  startTime: coreDataEntity.startTime,
+                  contribute: coreDataEntity.contribute,
+                  deviceId: coreDataEntity.deviceId,
+                  endTime: coreDataEntity.endTime,
+                  followedAt: coreDataEntity.followedAt,
+                  gotDeleted: coreDataEntity.gotDeleted,
+                  isIndoor: coreDataEntity.isIndoor,
+                  tags: coreDataEntity.tags,
+                  urlLocation: coreDataEntity.urlLocation,
+                  version: coreDataEntity.version,
+                  measurementStreams: (coreDataEntity.measurementStreams?.array as? [MeasurementStreamEntity])?.map(Database.MeasurementStream.init(coreDataEntity:)),
+                  status: coreDataEntity.status)
+    }
+}
+
+extension Database.MeasurementStream {
+    init(coreDataEntity: MeasurementStreamEntity) {
+        self.init(id: coreDataEntity.id,
+                  sensorName: coreDataEntity.sensorName,
+                  sensorPackageName: coreDataEntity.sensorPackageName,
+                  measurementType: coreDataEntity.measurementType,
+                  measurementShortType: coreDataEntity.measurementShortType,
+                  unitName: coreDataEntity.unitName,
+                  unitSymbol: coreDataEntity.unitSymbol,
+                  thresholdVeryHigh: coreDataEntity.thresholdVeryHigh,
+                  thresholdHigh: coreDataEntity.thresholdHigh,
+                  thresholdMedium: coreDataEntity.thresholdMedium,
+                  thresholdLow: coreDataEntity.thresholdLow,
+                  thresholdVeryLow: coreDataEntity.thresholdVeryLow)
+    }
+}
+
+extension URL {
+    init?(string: String?) {
```

Comment:
The type is preserved perfectly here, it will return `URL?` just like the `init?(string: String)` returns `URL?`. The only difference here is that this init here is also allowing us to pass optional strings in. I think it's pretty handy and doesn't violate any Swift ideas. But this extension was unused anyway 🤣 so I'll remove